pkg/softtainter/softtainter.go

```diff
@@ -0,0 +1,507 @@
+	var lnargs *nodeutilization.LowNodeUtilizationArgs
+
+	for _, p := range policy.Profiles {
+		if p.Name == string(desv1.LongLifecycle) || p.Name == string(desv1.RelieveAndMigrate) {
+			for _, pc := range p.PluginConfigs {
+				if pc.Name == nodeutilization.LowNodeUtilizationPluginName {
+					lnargs = pc.Args.(*nodeutilization.LowNodeUtilizationArgs)
+				}
+			}
+		}
+	}
+	if lnargs == nil {
+		err := fmt.Errorf("unable to read LowNodeUtilizationArgs")
+		logger.Error(err, "reconciliation failed")
+		return reconcile.Result{}, err
+	}
+
+	enableSoftTainter := false
+	if des.Spec.ProfileCustomizations != nil && des.Spec.ProfileCustomizations.DevEnableSoftTainter {
+		enableSoftTainter = true
+	}
+
+	st.args = &softTainterArgs{
+		useDeviationThresholds: lnargs.UseDeviationThresholds,
+		thresholds:             lnargs.Thresholds,
+		targetThresholds:       lnargs.TargetThresholds,
+		applySoftTaints:        enableSoftTainter,
+		mode:                   des.Spec.Mode,
+	}
+
+	if lnargs.MetricsUtilization == nil || lnargs.MetricsUtilization.Prometheus == nil {
+		err := fmt.Errorf("unable to read MetricsUtilization.Prometheus")
+		logger.Error(err, "reconciliation failed")
+		return reconcile.Result{}, err
+	}
+	st.promQuery = lnargs.MetricsUtilization.Prometheus.Query
+
+	var authToken *api.AuthToken
+	for _, provider := range policy.MetricsProviders {
+		if provider.Prometheus != nil {
+			st.promURL = provider.Prometheus.URL
+			authToken = provider.Prometheus.AuthToken
+		}
+	}
+	if st.promURL == "" {
+		err := fmt.Errorf("unable to read promURL")
+		logger.Error(err, "reconciliation failed")
+		return reconcile.Result{}, err
+	}
+
+	if authToken != nil {
+		err = st.reconcileSecretToken(ctx, authToken)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	} else {
+		err = st.reconcileInClusterSAToken()
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	st.resourceNames = getResourceNames(st.args.thresholds)
+
+	nodeSelector := labels.Everything()
+	if policy.NodeSelector != nil {
+		sel, err := labels.Parse(*policy.NodeSelector)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		nodeSelector = sel
+	}
+
+	nl := corev1.NodeList{}
+	err = st.client.List(ctx, &nl, &client.ListOptions{LabelSelector: nodeSelector})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	nodeList := make([]*corev1.Node, len(nl.Items))
+	for i, node := range nl.Items {
+		nodeList[i] = &node
+	}
+
+	if st.args.applySoftTaints {
+		err = st.syncSoftTaints(ctx, nodeList)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+	return reconcile.Result{RequeueAfter: st.resyncPeriod}, nil
+}
+
+func (st *softTainter) syncSoftTaints(ctx context.Context, nodes []*corev1.Node) error {
+	nodesMap := make(map[string]*corev1.Node, len(nodes))
+
+	nu := st.nodeUtilizationFactory(st.promClient, st.promQuery)
+	nodeUsages, err := nu.NodeUsageFromPrometheusMetrics(ctx)
+	if err != nil {
+		return err
+	}
+
+	// TODO: do we need this???
```

Comment:
You have a good question here. I guess that for the soft tainter with Prometheus the capacity will always be 100%.

Reply:
Likely yes, but I'm not 100% convinced because maybe in the future we can image other queries that could require something like this

Reply:
I guess we are OK with this then.

Reply:
I'll amend the comment to keep track of it